proposals/0723-qualified-strings.rst

```diff
@@ -0,0 +1,282 @@
+
+  * It's possible that this is one reason these extensions aren't defaults in GHC202X language editions, despite being in GHC for a long time.
+
+* Type inference ambiguity.
+
+  * Consider the following code:
+
+    ::
+
+      output :: IsString s => s -> IO ()
+
+      main = do
+        output "hello"
+        output "world"
+
+        -- output $ T.replace " " "_" input
+
+    This originally works with no extensions, due to the string literals being typed to concrete ``String``. But say the developer wants to add a call to ``T.replace`` and use ``Text`` literals; adding ``OverloadedStrings`` would cause ambiguity to the existing locations because they are now no longer concretely ``String``.
+
+This proposal would allow using a module qualifier to say precisely which function to desugar to, rather than using type classes, in a similar manner as ``-XQualifiedDo``. This would allow writing the previous code as
+
+::
+
+  {-# LANGUAGE QualifiedStrings #-}
+
+  main = do
+    output "hello"
+    output "world"
+
+    output $ T.replace T." " T."_" input
+
+The existing locations would continue working as ``String``, while the new line would unambiguously desugar to ``T.replace (T.pack " ") (T.pack "_") input``.
+
+Proposed Change Specification
+-----------------------------
+
+Introduce ``-XQualifiedStrings`` that desugars literal string syntax to function calls in a similar way to ``-XQualifiedDo`` (`docs <https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/qualified_do.html>`_, `proposal <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0216-qualified-do.rst>`_).
+
+As long as the desugared expressions/patterns type check, users are free to define these functions however they want. No whitespace is allowed between the ``.`` and the module name / literal.
+
+Currently, string literals have the following desugaring:
+
+.. list-table::
+    :align: left
+
+    * - **Expression**
+      - **Enabled extensions**
+      - **Desugared expression syntax**
+    * - ``"hello"``
+      -
+      - ``"hello"``
+    * - ``"hello"``
+      - ``-XOverloadedStrings``
+      - ``GHC.Exts.fromString "hello"``
+
+With ``-XQualifiedStrings``, we gain the following syntaxes:
+
+.. list-table::
+    :align: left
+
+    * - **New expression syntax**
+      - **Desugared expression syntax**
+    * - ``M."asdf"``
+      - ``M.fromString "asdf"``
+    * - ``M."""asdf"""``
+      - ``M.fromString "asdf"``
+
+.. list-table::
+    :align: left
+
+    * - **New pattern syntax**
+      - **Desugared pattern syntax**
+    * - ``M."asdf"``
+      - ``((== M.fromString "asdf") -> True)``
+    * - ``M."""asdf"""``
+      - ``((== M.fromString "asdf") -> True)``
+
+It is highly recommended that all types with ``IsString`` instances defined provide a ``.Interpolate`` module defining ``fromString = Data.String.fromString``, to enable locally-scoped overloading over ``-XOverloadedStrings``.
```

Comment:
I have no idea what a `.Interpolate` module is.  Am I supposed to know?

Perhaps you mean:

* It is highly recommended that, for every type `T` that has an `IsString` instance, the author of that instance also provides a module `MT` (the name is unimportant) defining a function
  ```
  fromString :: String -> T
  fromString = Data.String.fromString
  ```
  In this way a client of `T` is always free to choose to say `"foo"` (using `OverloadedStrings`) or `MT."foo"` (using `QualifiedStrings`).

Reply:
Yes, that's the crux of it. I'll reword

Reply:
Good idea, done